### PR TITLE
[builtins][arm64] Build __init_cpu_features_resolver on Apple platforms

### PR DESCRIPTION
**Pull request by jroelofs**

### CODE REVIEW

File: llvm/lib/Target/X86/X86AsmPrinter.cpp [hunk 1]
```diff
@@ -556,6 +556,8 @@ void X86AsmPrinter::emitGlobalIFunc(Module &M, const GlobalIFunc &GI) {
   JMP.setOpcode(X86::JMP_4);
   JMP.addOperand(MCOperand::createExpr(lowerConstant(GI.getResolver())));
   OutStreamer->emitInstruction(JMP, *Subtarget);
+
+  // FIXME: do the manual .symbol_resolver lowering that we did in AArch64AsmPrinter.
```
Comment (jroelofs) [inline llvm/lib/Target/X86/X86AsmPrinter.cpp:560]: - [ ] FIXME: This comment belongs in a different commit in the patch stack.

**Comment by jroelofs** [comment]
This is part of a patch series:
    * #73686
    * #73688
    * #73687
    * #73685

File: compiler-rt/lib/builtins/cpu_model.c [hunk 2]
```diff
@@ -1259,6 +1270,64 @@ struct {
...
+  int32_t val = 0;
+  size_t size = sizeof(val);
+  if (sysctlbyname(name, &val, &size, NULL, 0))
+    return false;
+  return val;
+}
+
+void __init_cpu_features_resolver(void) {
+  static dispatch_once_t onceToken = 0;
+  dispatch_once(&onceToken, ^{
```
Comment (MaskRay) [inline compiler-rt/lib/builtins/cpu_model.c:1289]: I am not familiar with blocks. Does this introduce any dependency on libcalls? Can this reuse `if (__aarch64_cpu_features.features)` used by ELF?

File: compiler-rt/lib/builtins/cpu_model.c [hunk 3]
```diff
@@ -1259,6 +1270,64 @@ struct {
...
+        {"hw.optional.arm.FEAT_LRCPC2", FEAT_RCPC2},
+        {"hw.optional.arm.FEAT_FRINTTS", FEAT_FRINTTS},
+        {"hw.optional.arm.FEAT_I8MM", FEAT_I8MM},
+        {"hw.optional.arm.FEAT_BF16", FEAT_BF16},
+        {"hw.optional.arm.FEAT_SB", FEAT_SB},
+        {"hw.optional.arm.FEAT_SSBS", FEAT_SSBS2},
+        {"hw.optional.arm.FEAT_BTI", FEAT_BTI},
+    };
+
+    for (size_t I = 0, E = sizeof(Features) / sizeof(Features[0]); I != E; ++I)
```
Comment (MaskRay) [inline compiler-rt/lib/builtins/cpu_model.c:1322]: compiler-rt generally uses lower-case variable names.

**Review by MaskRay — APPROVED** [review]

File: compiler-rt/lib/builtins/cpu_model.c (hunk 2 above)
Comment (jroelofs) [reply compiler-rt/lib/builtins/cpu_model.c:1289]: The only dependencies from the block + the dispatch_once should be on things in compiler-rt and in libSystem (which we need anyway for the sysctls). I am not sure whether we can get away with doing the `if (__aarch64_cpu_features.features)` check that ELF does, since the resolvers that will be calling this will happen lazily at runtime on the Darwin implementation, and therefore potentially by separate threads (unlike ELF, which would run them at load time). This function is idempotent, but I don't think that is enough of an atomicity guarantee, especially if `__aarch64_cpu_features` grows more fields.

Comment (MaskRay) [reply compiler-rt/lib/builtins/cpu_model.c:1289]: Thanks for the explanation. If `__init_cpu_features_resolver` can be called concurrently, I think it is worth a comment, since the other `__init_*` functions are guaranteed to be very early in the program startup sequence and is serial.

File: compiler-rt/lib/builtins/cpu_model.c [hunk 4]
```diff
@@ -1259,6 +1270,72 @@ struct {
...
+        {"hw.optional.arm.FEAT_FHM", FEAT_FP16FML},
+        {"hw.optional.arm.FEAT_DotProd", FEAT_DOTPROD},
+        {"hw.optional.arm.FEAT_RDM", FEAT_RDM},
+        {"hw.optional.arm.FEAT_LSE", FEAT_LSE},
+        {"hw.optional.floatingpoint", FEAT_FP},
+        {"hw.optional.AdvSIMD", FEAT_SIMD},
+        {"hw.optional.armv8_crc32", FEAT_CRC},
+        {"hw.optional.arm.FEAT_SHA1", FEAT_SHA1},
+        {"hw.optional.arm.FEAT_SHA256", FEAT_SHA2},
+        {"hw.optional.armv8_2_sha3", FEAT_SHA3},
```
Comment (ahmedbougacha) [inline compiler-rt/lib/builtins/cpu_model.c:1314]: I think you want `FEAT_SHA3` instead of `armv8_2_sha3` (and technically, sha512 also, but that shouldn't matter; crypto's a mess).  I thought we had one for crc32 as well, but looks like we don't.  I assume FP/SIMD are unneeded but harmless.

Otherwise, sysctl `FEAT_SPECRES` should map to enum `PREDRES`, and DIT, DPB/DPB2 are missing (DIT is vaguely plausibly useful, DPB not really.)  Rest LGTM!

**Review by ahmedbougacha — APPROVED** [review]
LG, ty

**Review by ilinpv — APPROVED** [review]
LGTM, thanks for adding Darwin support to cpu features detection!

**Comment by mstorsjo** [comment]
This commit broken building compiler-rt builtins for Windows on aarch64; building now hits these errors:
```
llvm-project/compiler-rt/lib/builtins/cpu_model.c:1192:2: error: No support for checking for lse atomics on this platfrom yet.
 1192 | #error No support for checking for lse atomics on this platfrom yet.
      |  ^
llvm-project/compiler-rt/lib/builtins/cpu_model.c:1571:2: error: No support for checking hwcap on this platform yet.
 1571 | #error No support for checking hwcap on this platform yet.
      |  ^
2 errors generated.
```
Before this change, most of this whole file was ifdeffed out when building on Windows (and Apple platforms, I would presume), but now most of it is included, then hitting this `#error`.

I guess it could work to just remove the `#error` cases, but this file suffers from a pretty deep ifdef nesting jungle, so I'm not sure if that's the best solution. (FWIW, if we wanted to add aarch64 CPU feature detection for Windows here, the code would be more of a separate codepath just like the Apple case, it doesn't share the linux/BSD HWCAP style.)

I can push a quick fix, either removing the `#error` or reverting this commit, later during the day.

BTW, when compiling the file I also get a bunch of warnings in this style:
```
llvm-project/compiler-rt/lib/builtins/cpu_model.c:1448:36: warning: value size does not match register size specified by the constraint and modifier [-Wasm-operand-widths]
 1448 |     getCPUFeature(ID_AA64PFR1_EL1, ftr);
      |                                    ^
llvm-project/compiler-rt/lib/builtins/cpu_model.c:1448:5: note: use constraint modifier "w"
 1448 |     getCPUFeature(ID_AA64PFR1_EL1, ftr);
      |     ^
llvm-project/compiler-rt/lib/builtins/cpu_model.c:1345:45: note: expanded from macro 'getCPUFeature'
 1345 | #define getCPUFeature(id, ftr) __asm__("mrs %0, " #id : "=r"(ftr))
      |
```

**Comment by DavidSpickett** [comment]
Also seen on Linaro's Windows on Arm 2 stage bot: https://lab.llvm.org/buildbot/#/builders/120/builds/5990

**Comment by jroelofs** [comment]
Sorry about that... thanks!

**Comment by jroelofs** [comment]
> BTW, when compiling the file I also get a bunch of warnings in this style:

@mstorsjo maybe `unsigned long` is 32 bits on that platform... what's the target triple?

**Comment by mstorsjo** [comment]
> > BTW, when compiling the file I also get a bunch of warnings in this style:
> 
> @mstorsjo maybe `unsigned long` is 32 bits on that platform... what's the target triple?

Ah, indeed - yes, Windows has 32 bit `long`s. The triples are `aarch64-windows-gnu` or `aarch64-windows-msvc`.

**Comment by jroelofs** [comment]
> this file suffers from a pretty deep ifdef nesting jungle, so I'm not sure if that's the best solution

I'll refactor, and put up a new PR.

**Comment by jroelofs** [comment]
https://github.com/llvm/llvm-project/pull/75635

**Comment by jroelofs** [comment]
Refactor here: https://github.com/llvm/llvm-project/pull/75635
